packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,413 @@
+    return [...this.warnings].flatMap((w) => this.createWarningStatement(w));
+  }
+
+  // TODO Rename this method
+  private findDeprecatedInTheTypeHierarchy(
+    toProcess: ts.Type[],
+  ): ts.Type | undefined {
+    if (toProcess.length === 0) {
+      return undefined;
+    }
+
+    const type = toProcess[0];
+    if (this.deprecatedTypes.includes(type)) {
+      return type;
+    }
+
+    const node = this.index.get(
+      this.typeChecker.getFullyQualifiedName(type.symbol),
+    );
+
+    // This type was not declared in the JSII assembly. Skipping.
+    if (node == null) {
+      return undefined;
+    }
+
+    if (!ts.isClassDeclaration(node) && !ts.isInterfaceDeclaration(node)) {
+      throw new Error('Node should be a class or interface declaration.');
+    }
+
+    const newBatch = toProcess.slice(1);
+    const heritageClauses = node.heritageClauses ?? [];
+    for (const clause of heritageClauses) {
+      for (const expression of clause.types) {
+        const type = this.typeChecker.getTypeAtLocation(expression);
+        if (!toProcess.includes(type)) {
+          newBatch.push(type);
+        }
+      }
+    }
+
+    return this.findDeprecatedInTheTypeHierarchy(newBatch);
+  }
+
+  private createWarningStatement(warning: Warning): ts.Statement[] {
+    const message = `${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`;
+    // TODO Are random numbers a good idea?
+    const functionName = `printWarning${getRandomInt(
+      0,
+      Number.MAX_SAFE_INTEGER,
+    )}`;
+
+    // TODO There must be a simpler way...
+    return [
+      // TODO Instead of declaring the function everywhere it's used, should we create a single declaration somewhere and reuse it?
+      //  If so, where should this declaration go?
+      ts.createFunctionDeclaration(
+        undefined,
+        undefined,
+        undefined,
+        functionName,
+        [],
+        [],
+        undefined,
+        ts.createBlock(
+          [
+            ts.createExpressionStatement(
+              ts.createAssignment(
+                ts.createIdentifier('const deprecated'),
+                ts.createIdentifier('process.env.DEPRECATED'),
+              ),
+            ),
+            ts.createExpressionStatement(
+              ts.createAssignment(
+                ts.createIdentifier('const deprecationMode'),
+                ts.createIdentifier(
+                  "['warn', 'error', 'quiet'].includes(deprecated) ? deprecated : \"warn\"",
+                ),
+              ),
+            ),
+            ts.createExpressionStatement(
+              ts.createAssignment(
+                ts.createIdentifier('const message'),
+                ts.createLiteral(message),
+              ),
+            ),
+            ts.createSwitch(
+              ts.createIdentifier('deprecationMode'),
+              ts.createCaseBlock([
+                ts.createCaseClause(ts.createLiteral('error'), [
```

Comment:
I'd call this mode `throw` or `fail` instead of `error`, because otherwise I'd have expected this outputs a message to STDERR.

Reply:
Makes sense, too.